torch/csrc/autograd/variable.cpp

```diff
@@ -167,27 +167,23 @@ void update_cpp_hooks_on_new_gradfn(
   const auto& meta = impl::get_autograd_meta(self);
   TORCH_INTERNAL_ASSERT(meta);
   TORCH_INTERNAL_ASSERT(new_fn);
-  if (!self.retains_grad()) {
-    // (1) reset the list when grad_fn is updated, so new hooks don't
-    //     get erroneously registered to the old grad_fn.
-    //     Note that the old cpp_hooks_list_ is still kept alive by the
-    //     old grad_fn so hooks registered to the older version of the tensor
-    //     will continue to be active.
-    meta->cpp_hooks_list_ = nullptr;
-    return;
-  }
+  // (1) reset the list when grad_fn is updated, so new hooks don't
+  //     get erroneously registered to the old grad_fn.
+  //     Note that the old cpp_hooks_list_ is still kept alive by the
+  //     old grad_fn so hooks registered to the older version of the tensor
+  //     will continue to be active.
+  meta->cpp_hooks_list_ = nullptr;
   // (2) If there is a retains_grad hook registered, move that from the
   //     old cpp_hooks_list_ to the new one
-  auto idx = meta->retains_grad_;
-  auto new_list = std::make_shared<hooks_list>();
-  new_list->push_back(std::move((*meta->cpp_hooks_list_)[idx]));
-  (*meta->cpp_hooks_list_)[idx] = nullptr;
-  meta->cpp_hooks_list_ = new_list;
-  // Since this is a new list, 0 is the index of the retains_grad hook
-  meta->retains_grad_ = 0;
-  std::unique_ptr<FunctionPreHook> hook_ptr(
-      new CppFunctionPreHook(meta->cpp_hooks_list_, self.output_nr()));
-  new_fn->add_pre_hook(std::move(hook_ptr));
+  if (self.retains_grad()) {
+    auto new_list = std::make_shared<hooks_list>();
```

Comment:
This is quite confusing :/
Do you think making the retain_grad hook field not a vector would simplify things?

Reply:
Yes, I was planning on doing that in a follow up, but I can also update that in this PR.

Reply:
Since this is the way it was before, are you okay with keeping this for now, and leaving that as a follow up. If not, let me know.

Reply:
Sure, could you open a BE issue so that we don't forget?